(closes #2896) Adds IncreaseRankLoopArraysTrans

sergisiso opened the pull request.

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -0,0 +1,130 @@
+def test_trans_name_and_str():
+    ''' Check that the name and str method works as expected. '''
+    trans = IncreaseRankLoopArraysTrans()
+    assert trans.name == "IncreaseRankLoopArraysTrans"
```
You don't need this as `.name` just calls the base-class implementation.

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -0,0 +1,130 @@
+        program test
+            real, dimension(10) :: a
+            integer :: i = 3
+            a(i) = 3
```
`validate` should always fail for this code as we won't touch the reference of `a` in this assignment?

arporter replied:
Please add tests for accesses before and after the target loop. Alternatively, as mentioned previously, perhaps this transformation should target a whole scoping region in which the array is local?

sergisiso replied:
I provided a different solution, let me know what you think. In this case this example should pass the validate.

arporter replied:
I quite like the new solution but, as commented earlier, I think it needs to be made a bit more robust.

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -0,0 +1,130 @@
+            "referenced inside a Codeblock, but 'a' is inside one."
+            in str(err.value))
+
+    # Referencing non-existan arrays is fine, they are just ignored, this
```
Please add this information to the docstring of the transformation.

sergisiso replied:
This is not the case anymore

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -0,0 +1,130 @@
+             end do
+             do j = 1, N
+                 ztmp(j) = ztmp(j) + 1
+             end do
```
What would happen if we had e.g. `ztmp(:) = 0` somewhere inside the `i` loop?

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -0,0 +1,130 @@
+    code = fortran_writer(psyir)
+
+    # The declaration and references have been updated
+    assert "real, dimension(n,-5:m + 3) :: ztmp" in code
```
We need a separate test for a case where e.g. `m` is written to before the `i` loop.

sergisiso replied:
I think one of the new tests covers this

arporter reviewed (changes requested):
Thanks Sergi. I have some concerns about the new transformation as it stands - it will be very easy to make it generate code that does not compile. We perhaps need to discuss whether it should target a whole scoping region rather than a single loop? Either way, it needs some extra validation checks and associated tests.

We will also need to update the User Guide with information on this transformation.

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,197 @@
+        return ("Increases the Rank of the supplied arrays by the iteration "
+                "space of the given loop, and update all its references")
+
+    def validate(self, node, options=None):
```
Done

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,197 @@
+
+        :raises TransformationError: if the node is not a Loop.
+        :raises TransformationError: if the node is not inside a Routine.
+        :raises TransformationError: the array option has not been provided.
```
Done

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,197 @@
+                    f"{self.name} does not support arrays that are referenced "
+                    f"inside a Codeblock, but '{array.name}' is inside one.")
+
+            # TODO: What should happen with references outside the loop?
```
Applying it to a loop is very convinient because it also provide the iteration space (size of the new rank) and iteration variable (to use as indexing value inside the loop). I will instead add a new loop around accesses outside the selected loop (the NEMO this are mostly init statement e.g `ztmp(:) = 0`, which now need to become `ztmp(:,:) = 0` - but with explicit loops for adding directives).

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,197 @@
+                f"The target of the {self.name} transformation should be a "
+                f"Loop inside a Routine.")
+
+        # Capture all symbols used inside codeblocks, this are not permited
```
Fixed

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,197 @@
+        for array in array_list:
+            if isinstance(array, str):
+                array = node.scope.symbol_table.lookup(array, otherwise=None)
+            if array is None:
```
I agree with this. If a name is given it must exist. But I will provide a second argument for more oriented to searching patterns (e.g. symbols that start with 'z')

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,197 @@
+            if array is None:
+                continue
+
+            # TODO: check that the bound expressions are valid as static
```
Done (I don't think we can currently guarantee this, but I tried a best-effort check)

sergisiso commented on src/psyclone/psyir/transformations/omp_target_trans.py:
```diff
@@ -184,6 +190,24 @@ def validate(self, node, options=None):
+                            )
+                            if verbose:
+                                node.preceding_comment = message
+                            raise TransformationError(message)
```
Done

sergisiso commented on src/psyclone/psyir/transformations/omp_target_trans.py:
```diff
@@ -161,18 +162,23 @@ def validate(self, node, options=None):
             attempts to enclose the assingment setting the return value.
         '''
         device_string = options.get("device_string", "") if options else ""
+        strings = options.get("allow_strings", False) if options else False
```
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
+                f"inside a Routine.")
+
+        # Check if the loop bound expressions are static
+        values_to_check = set()
```
Comment pls, e.g. "First, find all Symbols accessed in the loop bounds."

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
+                if ref.parent.is_inquiry or ref.position == 0:
+                    continue
+            values_to_check.add(ref.symbol)
+        for assignment in routine.walk(Assignment):
```
"# Second, check that none of these Symbols are assigned to within the Routine."

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
+                if assignment.lhs.symbol in values_to_check:
+                    raise TransformationError(
+                        f"{self.name} can only be applied to loops with static"
+                        f" loop bound expressions, but in has been attempted "
```
s/in/it/

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
+                        f"{self.name} can only be applied to loops with static"
+                        f" loop bound expressions, but in has been attempted "
+                        f"in a loop with the variable "
+                        f"'{assignment.lhs.symbol.name}'."
```
...' which is assigned to: {assignment.debug_string().strip()}'."

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
-        if len(array_list) == 0:
+        if not isinstance(arrays, list) or len(arrays) == 0:
             raise TransformationError(
                 f"{self.name} has a mandatory 'arrays' option that "
```
"...that is required to specify which arrays are to have their rank increased."

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
                 raise TransformationError(
                     f"{self.name} does not support arrays that are referenced "
                     f"inside a Codeblock, but '{array.name}' is inside one.")
+            # TODO: check that the bound expressions are valid as static
```
This one is done now?

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -99,28 +98,30 @@ class IncreaseRankLoopArraysTrans(Transformation):
-        if options is None:
-            options = {}
-
         if not isinstance(node, Loop):
```
I'm still getting to grips with this myself but I *think* validate should just have the **kwargs argument and then call `get_option` and `validate_options` - see https://psyclone.readthedocs.io/en/latest/developer_guide/transformations.html#moving-to-the-new-transformations-options

arporter replied:
I'm happy to keep the specific checks that you already have but I think we still then need a following call to `validate_options` since that checks for any unsupported options supplied by the caller. Is that correct @LonelyCat124 ? Also, we need to change the signature of `validate` so that it just has **kwargs. So, I think we'd call `get_option` to get the "arrays" option, then do our specific checks with explicit error messages, then call `validate_options`.

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
-        self.validate(node, options)
-
-        array_list = options.get("arrays", [])
+        self.validate(node, arrays, **kwargs)
```
Probably this needs to be `arrays=arrays`?

arporter replied:
I think it's safer in case other arguments get added in future (unlikely in this case). I see you've made the change anyway so thanks :-)

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
 
-            # TODO: check that the bound expressions are valid as static
-            # expressions in the declarations.
             array.shape.append(
```
Comment pls - "Add an additional dimension to this array with the same bounds as the target Loop"

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -33,26 +33,25 @@
+    size of the interation space. Then it indexes all accesses with the loop
+    variable, so that each iteration accesses a unique location. Effectively
+    making the sub-array private for each iteration of the loop.
 
```
Please also document the fact that it modifies all accesses outside the target Loop too. Please could you extend the example in the docstring to show this.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
                 if ref.symbol is array:
-                    ref.addchild(Reference(node.variable))
+                    if ref.is_descendent_of(node):
+                        # Inside the target loop index the reference to the
```
s/to/using/

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -129,69 +130,101 @@ def validate(self, node, options=None):
+                        # loop variable
+                        ref.addchild(Reference(node.variable))
+                    else:
+                        # Outside the target loop index the whole range
```
What about if the scoping region contains two explicit loops that both use the same array (before its rank is increased)? I think this will end up transforming the accesses inside a second loop to use full slices. We should probably catch and reject this case in `validate`.

sergisiso replied:
> I think this will end up transforming the accesses inside a second loop to use full slices.
Yes it will. But I think this is the right behaviour, see last example with the initialisation test. We essentially trade more parallelism (independent elements), by duplicating this elements outside the loop.

sergisiso replied:
Ok, I partially came around on this since our discussion and now I am much more restrictive in that the validate only allows simple initialisation expressions "ztmp(..) = ..." outside the loop. As evaluating the correctness of other cases was very complex. I think with this I don't need the warning in the docstring, as now I am more confident that the semantics do not change. And the previous condition already blocks applying it two times, as it will find the array usage in the other loop unless they are simple write-only loops, in which case is also semantically equivalent. Let me know what you think.

arporter commented on src/psyclone/psyir/transformations/omp_target_trans.py:
```diff
@@ -147,11 +147,17 @@ def validate(self, node, options=None):
         :type options: Optional[Dict[str, Any]]
         :param str options["device_string"]: provide a compiler-platform
             identifier.
+        :param str options["allow_strings"]: prermit OMP target regions
```
"permit"

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -80,25 +79,63 @@ def test_irla_validate(fortran_reader):
+            "in a loop with the variable 'variable'." in str(err.value))
+
+    # This is fine
+    trans.apply(routine.walk(Loop)[2], arrays=['a'])
```
Actually, I suspect it isn't - please check the output.

sergisiso replied:
Oh did you mean because the ArrayBound missed the dangling_parent to avoid the "linebreak" in the declaration?

sergisiso replied:
> So the first loop now writes to a(i,:) every trip which isn't great.

@arporter The example may be a bit unfortunate but the original first loop already writes to `a(i)` multiple time:
```
            do j = 1, variable
                do i = 1, 10
                   a(i) = 3
                end do
            enddo
```
so the transformed code just does the same but for the whole slice `a(i,:)`

arporter commented on src/psyclone/tests/psyir/transformations/increase_rank_loop_arrays_test.py:
```diff
@@ -121,10 +158,47 @@ def test_irla_apply(fortran_reader, fortran_writer):
+    assert "ztmp(j,:) = ztmp(j,:) / 3" in code
+    # And the two inside the loop
+    assert "ztmp(j,i) = ztmp(j,i) + 4" in code
+    assert "ztmp(:,i) = 5" in code
```
Please add a compilation check.

sergisiso replied:
Done

arporter reviewed (changes requested):
Thanks Sergi, I'm a lot happier now. I like the new direction although I think some of the validation needs a bit more tightening. I think the use of `**kwargs` also needs a bit of work.

sergisiso commented:
@arporter This is also ready for another look

arporter reviewed (changes requested):
Thanks Sergi. I think we need to agree what we want this transformation to do if there's more than one loop containing assignments to an array that we wish to increase the rank of.
Apart from that, we also need to agree with @LonelyCat124 what we should be doing with `validate`.
Finally (probably after we've reached agreement), the User Guide needs updating so that it mentions this transformation.

arporter commented:
The only potential improvement I can think of is to perhaps create a new, temporary array rather than extend the size of the existing one. That way, we could copy (in parallel) into it before the loop we are transforming and leave all other uses alone. This assumes the array doesn't have a DUC that extends outside of the target loop but perhaps that's already a condition of the transformation?

sergisiso commented:
@arporter This is ready for another look, instead of the proposed scattering to a new temporary, I made the current implementation much more restrictive and only allow simple 'write-only' expressions outside the loop, which effectively makes it not possible to apply to two loops that "use" the arrays, as each will complain about the other one. It also prevents rank-changes of expressions outside the loop which would have created invalid code. Let me know what you think of this.

arporter commented:
Another question, is the small performance improvement:
<img width="843" height="217" alt="image" src="https://github.com/user-attachments/assets/9bd7eb0b-4e5e-4c7f-858f-7f27a2cdc8c1" />
as expected? (I thought it would be larger.)

sergisiso commented:
> Another question, is the small performance improvement: as expected? (I thought it would be larger.)

To comment on this, the loops that this optimisation targets were already significanly improved by the array_privatisation option I added some time ago. That transformation already allowed the outer loop to be offloaded and avoided the data traffic which was the biggest performance penalty.

The IncreaseRankLoopArraysTrans by itself just does the same, with approx the same perf results. The advantage of this new transformation is that it can be followed up by bringing the outer-loop next to the inner loop (done in utils.py), which allows to collapse them and have more parallelism to distribute the loop. This is the benefit that you see in the chart above but is smaller than avoiding the data copies that we already had.

In the nemo sct it was a big win because they got both benefits at once.

There is a second benefit (again not for the transformation itself but by enabling the follow up loop reordering), which is that moving outer-loop down makes the condition which sometimes have characters end up outside the offloaded region. For some reason the compiler was fine with this string comparisons on the GPU, but I am more confident with them being outside.

sergisiso commented on examples/nemo/scripts/utils.py:
```diff
@@ -332,10 +334,12 @@ def normalise_loops(
         where possible, default is False.
+    :param increase_array_ranks: whether to increase the rank of selected
+        arrays.
+    :param bool hoist_expressions: whether to hoist bounds and loop invariant
```
Deleted

sergisiso commented on examples/nemo/scripts/utils.py:
```diff
@@ -410,6 +417,62 @@ def normalise_loops(
 
 
+def increase_rank_and_reorder_nemov5_loops(routine: Routine):
+    ''' This method increases the rank of temporary loops used inside selected
```
Done

sergisiso commented on src/psyclone/psyir/transformations/increase_rank_loop_arrays_trans.py:
```diff
@@ -0,0 +1,266 @@
+        '''Applies the transformation.
+
+        :param node: target Loop node.
+        :param arrays: list of arrays that will have the rank increased.
```
Done

sergisiso commented:
@arporter This is ready for another look

> When you say "simple initialisation expressions" did you mean "the target array is only assigned to and not read from"? If you could add a couple of tests for cases that validate() should reject that would help me. Probably also good to add this to (perhaps a second) example in the docstring of the class.

I see that using "initialisation" was confusing, I have updated the text regarding this and added a new validation test. I also modified the docstring example to be more explicit about this, however it is difficult to add a second (essentially different) docsting example because this is pretty much all that is premitted now.
